Kafka message payload types are not registered for reflection

## Describe the bug
Serde message payloads need to be registered for reflection in native builds.
If these classes were not registered by another means (i.e resteasy extension) message serde does not work in native mode.

### Expected behavior

Message serde (jackson/jsonb) works in native mode.

### Actual behavior

For example with Jackson serialization fails with a similar stacktrace:

```
Caused by: java.lang.RuntimeException: com.fasterxml.jackson.databind.exc.InvalidDefinitionException: No serializer found for class org.acme.kafka.model.Quote and no properties discovered to create BeanSerializer (to avoid exception, disable SerializationFeature.FAIL_ON_EMPTY_BEANS)
	at io.quarkus.kafka.client.serialization.ObjectMapperSerializer.serialize(ObjectMapperSerializer.java:37)
	at org.apache.kafka.common.serialization.Serializer.serialize(Serializer.java:62)
	at org.apache.kafka.clients.producer.KafkaProducer.doSend(KafkaProducer.java:925)
	at org.apache.kafka.clients.producer.KafkaProducer.send(KafkaProducer.java:885)
	at io.smallrye.reactive.messaging.kafka.impl.ReactiveKafkaProducer.lambda$send$3(ReactiveKafkaProducer.java:96)
	at io.smallrye.context.impl.wrappers.SlowContextualConsumer.accept(SlowContextualConsumer.java:21)
	at io.smallrye.mutiny.operators.uni.builders.UniCreateWithEmitter.subscribe(UniCreateWithEmitter.java:22)
	... 7 more
Caused by: com.fasterxml.jackson.databind.exc.InvalidDefinitionException: No serializer found for class org.acme.kafka.model.Quote and no properties discovered to create BeanSerializer (to avoid exception, disable SerializationFeature.FAIL_ON_EMPTY_BEANS)
	at com.fasterxml.jackson.databind.SerializerProvider.reportBadDefinition(SerializerProvider.java:1276)
	at com.fasterxml.jackson.databind.DatabindContext.reportBadDefinition(DatabindContext.java:400)
	at com.fasterxml.jackson.databind.ser.impl.UnknownSerializer.failForEmpty(UnknownSerializer.java:71)
	at com.fasterxml.jackson.databind.ser.impl.UnknownSerializer.serialize(UnknownSerializer.java:33)
	at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider._serialize(DefaultSerializerProvider.java:480)
	at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider.serializeValue(DefaultSerializerProvider.java:319)
	at com.fasterxml.jackson.databind.ObjectMapper._writeValueAndClose(ObjectMapper.java:4487)
	at com.fasterxml.jackson.databind.ObjectMapper.writeValue(ObjectMapper.java:3701)
	at io.quarkus.kafka.client.serialization.ObjectMapperSerializer.serialize(ObjectMapperSerializer.java:34)
	... 13 more
``` 

## To Reproduce

Steps to reproduce the behavior:
1. Setup app with `smallrye-reactive-kafka` extension
2. Send/receive serde'd object to/from Kafka
3. Make sure the object is not already registered for reflection by another extension, (i.e REST endpoint payload)
4. Run the app in native mode

--- a/extensions/smallrye-reactive-messaging-kafka/deployment/src/main/java/io/quarkus/smallrye/reactivemessaging/kafka/deployment/SmallRyeReactiveMessagingKafkaProcessor.java
+++ b/extensions/smallrye-reactive-messaging-kafka/deployment/src/main/java/io/quarkus/smallrye/reactivemessaging/kafka/deployment/SmallRyeReactiveMessagingKafkaProcessor.java
@@ -95,17 +95,9 @@ public class SmallRyeReactiveMessagingKafkaProcessor {
                 continue;
             }
 
-            Type injectionPointType;
-            switch (annotation.target().kind()) {
-                case FIELD:
-                    injectionPointType = annotation.target().asField().type();
-                    break;
-                case METHOD_PARAMETER:
-                    MethodParameterInfo parameter = annotation.target().asMethodParameter();
-                    injectionPointType = parameter.method().parameters().get(parameter.position());
-                    break;
-                default:
-                    continue;
+            Type injectionPointType = getInjectionPointTypeFromChannel(annotation);
+            if (injectionPointType == null) {
+                continue;
             }
 
             processIncomingChannelInjectionPoint(discovery, injectionPointType, (keyDeserializer, valueDeserializer) -> {
@@ -124,6 +116,18 @@ public class SmallRyeReactiveMessagingKafkaProcessor {
         }
     }
 
+    private Type getInjectionPointTypeFromChannel(AnnotationInstance annotation) {
+        switch (annotation.target().kind()) {
+            case FIELD:
+                return annotation.target().asField().type();
+            case METHOD_PARAMETER:
+                MethodParameterInfo parameter = annotation.target().asMethodParameter();
+                return parameter.method().parameters().get(parameter.position());
+            default:
+                return null;
+        }
+    }
+
     void produceRuntimeConfigurationDefaultBuildItem(DefaultSerdeDiscoveryState discovery,
             BuildProducer<RunTimeConfigurationDefaultBuildItem> config, String key, String value) {
         discovery.runIfConfigIsAbsent(key, value,
@@ -132,6 +136,10 @@ public class SmallRyeReactiveMessagingKafkaProcessor {
 
     private void processIncomingMethod(DefaultSerdeDiscoveryState discovery, MethodInfo method,
             BiConsumer<String, String> deserializerAcceptor) {
+        processIncomingType(discovery, getIncomingType(method), deserializerAcceptor);
+    }
+
+    private Type getIncomingType(MethodInfo method) {
         List<Type> parameterTypes = method.parameters();
         int parametersCount = parameterTypes.size();
         Type returnType = method.returnType();
@@ -169,44 +177,37 @@ public class SmallRyeReactiveMessagingKafkaProcessor {
                 incomingType = incomingType.asParameterizedType().arguments().get(0);
             }
         }
-
-        processIncomingType(discovery, incomingType, deserializerAcceptor);
+        return incomingType;
     }
 
     private void processIncomingChannelInjectionPoint(DefaultSerdeDiscoveryState discovery, Type injectionPointType,
             BiConsumer<String, String> deserializerAcceptor) {
-        Type incomingType = null;
+        processIncomingType(discovery, getIncomingChannelType(injectionPointType), deserializerAcceptor);
+    }
 
-        if (isPublisher(injectionPointType) || isPublisherBuilder(injectionPointType) || isMulti(injectionPointType)) {
-            incomingType = injectionPointType.asParameterizedType().arguments().get(0);
+    private Type getIncomingChannelType(Type injectionPointType) {
+        if (injectionPointType == null) {
+            return null;
         }
-
-        processIncomingType(discovery, incomingType, deserializerAcceptor);
+        if (isPublisher(injectionPointType) || isPublisherBuilder(injectionPointType) || isMulti(injectionPointType)) {
+            return injectionPointType.asParameterizedType().arguments().get(0);
+        } else {
+            return null;
+        }
     }
 
     private void processIncomingType(DefaultSerdeDiscoveryState discovery, Type incomingType,
             BiConsumer<String, String> deserializerAcceptor) {
-        if (incomingType == null) {
-            return;
-        }
-
-        if (isMessage(incomingType)) {
-            List<Type> typeArguments = incomingType.asParameterizedType().arguments();
-            String deserializer = deserializerFor(discovery, typeArguments.get(0));
-            deserializerAcceptor.accept(null, deserializer);
-        } else if (isKafkaRecord(incomingType) || isRecord(incomingType) || isConsumerRecord(incomingType)) {
-            List<Type> typeArguments = incomingType.asParameterizedType().arguments();
-            String keyDeserializer = deserializerFor(discovery, typeArguments.get(0));
-            String valueDeserializer = deserializerFor(discovery, typeArguments.get(1));
-            deserializerAcceptor.accept(keyDeserializer, valueDeserializer);
-        } else if (isRawMessage(incomingType)) {
-            String deserializer = deserializerFor(discovery, incomingType);
-            deserializerAcceptor.accept(null, deserializer);
-        }
+        extractKeyValueType(incomingType, (key, value) ->
+                deserializerAcceptor.accept(deserializerFor(discovery, key), deserializerFor(discovery, value)));
     }
 
     private void processOutgoingMethod(DefaultSerdeDiscoveryState discovery, MethodInfo method,
             BiConsumer<String, String> serializerAcceptor) {
+        processOutgoingType(discovery, getOutgoingType(method), serializerAcceptor);
+    }
+
+    private Type getOutgoingType(MethodInfo method) {
         List<Type> parameterTypes = method.parameters();
         int parametersCount = parameterTypes.size();
         Type returnType = method.returnType();
@@ -245,39 +246,44 @@ public class SmallRyeReactiveMessagingKafkaProcessor {
                 outgoingType = outgoingType.asParameterizedType().arguments().get(0);
             }
         }
-
-        processOutgoingType(discovery, outgoingType, serializerAcceptor);
+        return outgoingType;
     }
 
     private void processOutgoingChannelInjectionPoint(DefaultSerdeDiscoveryState discovery, Type injectionPointType,
             BiConsumer<String, String> serializerAcceptor) {
-        Type outgoingType = null;
+        processOutgoingType(discovery, getOutgoingChannelType(injectionPointType), serializerAcceptor);
+    }
 
-        if (isEmitter(injectionPointType) || isMutinyEmitter(injectionPointType)) {
-            outgoingType = injectionPointType.asParameterizedType().arguments().get(0);
+    private Type getOutgoingChannelType(Type injectionPointType) {
+        if (injectionPointType == null) {
+            return null;
         }
-
-        processOutgoingType(discovery, outgoingType, serializerAcceptor);
+        if (isEmitter(injectionPointType) || isMutinyEmitter(injectionPointType)) {
+            return injectionPointType.asParameterizedType().arguments().get(0);
+        } else {
+            return null;
+        }
     }
 
     private void processOutgoingType(DefaultSerdeDiscoveryState discovery, Type outgoingType,
             BiConsumer<String, String> serializerAcceptor) {
-        if (outgoingType == null) {
+        extractKeyValueType(outgoingType, (key, value) ->
+                serializerAcceptor.accept(serializerFor(discovery, key), serializerFor(discovery, value)));
+    }
+
+    private void extractKeyValueType(Type type, BiConsumer<Type, Type> keyValueTypeAcceptor) {
+        if (type == null) {
             return;
         }
 
-        if (isMessage(outgoingType)) {
-            List<Type> typeArguments = outgoingType.asParameterizedType().arguments();
-            String serializer = serializerFor(discovery, typeArguments.get(0));
-            serializerAcceptor.accept(null, serializer);
-        } else if (isKafkaRecord(outgoingType) || isRecord(outgoingType) || isProducerRecord(outgoingType)) {
-            List<Type> typeArguments = outgoingType.asParameterizedType().arguments();
-            String keySerializer = serializerFor(discovery, typeArguments.get(0));
-            String valueSerializer = serializerFor(discovery, typeArguments.get(1));
-            serializerAcceptor.accept(keySerializer, valueSerializer);
-        } else if (isRawMessage(outgoingType)) {
-            String serializer = serializerFor(discovery, outgoingType);
-            serializerAcceptor.accept(null, serializer);
+        if (isMessage(type)) {
+            List<Type> typeArguments = type.asParameterizedType().arguments();
+            keyValueTypeAcceptor.accept(null, typeArguments.get(0));
+        } else if (isKafkaRecord(type) || isRecord(type) || isProducerRecord(type) || isConsumerRecord(type)) {
+            List<Type> typeArguments = type.asParameterizedType().arguments();
+            keyValueTypeAcceptor.accept(typeArguments.get(0), typeArguments.get(1));
+        } else if (isRawMessage(type)) {
+            keyValueTypeAcceptor.accept(null, type);
         }
     }
 
@@ -485,6 +491,9 @@ public class SmallRyeReactiveMessagingKafkaProcessor {
     }
 
     private String serializerDeserializerFor(DefaultSerdeDiscoveryState discovery, Type type, boolean serializer) {
+        if (type == null) {
+            return null;
+        }
         DotName typeName = type.name();
 
         // statically known serializer/deserializer

--- a/extensions/smallrye-reactive-messaging-kafka/deployment/src/main/java/io/quarkus/smallrye/reactivemessaging/kafka/deployment/SmallRyeReactiveMessagingKafkaProcessor.java
+++ b/extensions/smallrye-reactive-messaging-kafka/deployment/src/main/java/io/quarkus/smallrye/reactivemessaging/kafka/deployment/SmallRyeReactiveMessagingKafkaProcessor.java
@@ -3,10 +3,15 @@ package io.quarkus.smallrye.reactivemessaging.kafka.deployment;
 import java.util.List;
 import java.util.Map;
 import java.util.function.BiConsumer;
+import java.util.function.Function;
 
+import org.eclipse.microprofile.config.Config;
+import org.eclipse.microprofile.config.ConfigProvider;
+import org.eclipse.microprofile.config.ConfigValue;
 import org.jboss.jandex.AnnotationInstance;
 import org.jboss.jandex.ClassInfo;
 import org.jboss.jandex.DotName;
+import org.jboss.jandex.IndexView;
 import org.jboss.jandex.MethodInfo;
 import org.jboss.jandex.MethodParameterInfo;
 import org.jboss.jandex.Type;
@@ -14,9 +19,11 @@ import org.jboss.jandex.Type;
 import io.quarkus.deployment.Feature;
 import io.quarkus.deployment.annotations.BuildProducer;
 import io.quarkus.deployment.annotations.BuildStep;
+import io.quarkus.deployment.annotations.Consume;
 import io.quarkus.deployment.builditem.CombinedIndexBuildItem;
 import io.quarkus.deployment.builditem.FeatureBuildItem;
 import io.quarkus.deployment.builditem.RunTimeConfigurationDefaultBuildItem;
+import io.quarkus.deployment.builditem.RuntimeConfigSetupCompleteBuildItem;
 import io.quarkus.deployment.builditem.nativeimage.ReflectiveClassBuildItem;
 import io.vertx.kafka.client.consumer.impl.KafkaReadStreamImpl;
 
@@ -198,8 +205,8 @@ public class SmallRyeReactiveMessagingKafkaProcessor {
 
     private void processIncomingType(DefaultSerdeDiscoveryState discovery, Type incomingType,
             BiConsumer<String, String> deserializerAcceptor) {
-        extractKeyValueType(incomingType, (key, value) ->
-                deserializerAcceptor.accept(deserializerFor(discovery, key), deserializerFor(discovery, value)));
+        extractKeyValueType(incomingType, (key, value) -> deserializerAcceptor.accept(deserializerFor(discovery, key),
+                deserializerFor(discovery, value)));
     }
 
     private void processOutgoingMethod(DefaultSerdeDiscoveryState discovery, MethodInfo method,
@@ -267,8 +274,8 @@ public class SmallRyeReactiveMessagingKafkaProcessor {
 
     private void processOutgoingType(DefaultSerdeDiscoveryState discovery, Type outgoingType,
             BiConsumer<String, String> serializerAcceptor) {
-        extractKeyValueType(outgoingType, (key, value) ->
-                serializerAcceptor.accept(serializerFor(discovery, key), serializerFor(discovery, value)));
+        extractKeyValueType(outgoingType,
+                (key, value) -> serializerAcceptor.accept(serializerFor(discovery, key), serializerFor(discovery, value)));
     }
 
     private void extractKeyValueType(Type type, BiConsumer<Type, Type> keyValueTypeAcceptor) {
@@ -540,5 +547,112 @@ public class SmallRyeReactiveMessagingKafkaProcessor {
 
         // unknown
         return null;
+    }
+
+    @BuildStep
+    @Consume(RuntimeConfigSetupCompleteBuildItem.class)
+    public void reflectiveValueSerializerPayload(CombinedIndexBuildItem combinedIndex,
+            BuildProducer<ReflectiveClassBuildItem> reflectiveClass) {
+        IndexView index = combinedIndex.getIndex();
+        Config config = ConfigProvider.getConfig();
+
+        processOutgoingForReflectiveClassPayload(index, config,
+                (annotation, payloadType) -> produceReflectiveClass(reflectiveClass, payloadType));
+
+        processOutgoingChannelForReflectiveClassPayload(index, config,
+                (annotation, payloadType) -> produceReflectiveClass(reflectiveClass, payloadType));
+
+        processIncomingForReflectiveClassPayload(index, config,
+                (annotation, payloadType) -> produceReflectiveClass(reflectiveClass, payloadType));
+
+        processIncomingChannelForReflectiveClassPayload(index, config,
+                (annotation, payloadType) -> produceReflectiveClass(reflectiveClass, payloadType));
+
+    }
+
+    void produceReflectiveClass(BuildProducer<ReflectiveClassBuildItem> reflectiveClass, Type type) {
+        reflectiveClass.produce(new ReflectiveClassBuildItem(true, true, true, type.name().toString()));
+    }
+
+    // visible for testing
+    void processOutgoingForReflectiveClassPayload(IndexView index, Config config,
+            BiConsumer<AnnotationInstance, Type> annotationAcceptor) {
+        processAnnotationsForReflectiveClassPayload(index, config, DotNames.OUTGOING, true,
+                annotation -> getOutgoingType(annotation.target().asMethod()), annotationAcceptor);
+    }
+
+    // visible for testing
+    void processOutgoingChannelForReflectiveClassPayload(IndexView index, Config config,
+            BiConsumer<AnnotationInstance, Type> annotationAcceptor) {
+        processAnnotationsForReflectiveClassPayload(index, config, DotNames.CHANNEL, true,
+                annotation -> getOutgoingChannelType(getInjectionPointTypeFromChannel(annotation)), annotationAcceptor);
+    }
+
+    // visible for testing
+    void processIncomingForReflectiveClassPayload(IndexView index, Config config,
+            BiConsumer<AnnotationInstance, Type> annotationAcceptor) {
+        processAnnotationsForReflectiveClassPayload(index, config, DotNames.INCOMING, false,
+                annotation -> getIncomingType(annotation.target().asMethod()), annotationAcceptor);
+    }
+
+    // visible for testing
+    void processIncomingChannelForReflectiveClassPayload(IndexView index, Config config,
+            BiConsumer<AnnotationInstance, Type> annotationAcceptor) {
+        processAnnotationsForReflectiveClassPayload(index, config, DotNames.CHANNEL, false,
+                annotation -> getIncomingChannelType(getInjectionPointTypeFromChannel(annotation)),
+                annotationAcceptor);
+    }
+
+    private void processAnnotationsForReflectiveClassPayload(IndexView index, Config config, DotName annotationType,
+            boolean serializer, Function<AnnotationInstance, Type> typeExtractor,
+            BiConsumer<AnnotationInstance, Type> annotationAcceptor) {
+        for (AnnotationInstance annotation : index.getAnnotations(annotationType)) {
+            String channelName = annotation.value().asString();
+            Type type = typeExtractor.apply(annotation);
+            extractKeyValueType(type, (key, value) -> {
+                if (key != null && isSerdeJson(index, config, channelName, serializer, true)) {
+                    annotationAcceptor.accept(annotation, key);
+                }
+                if (value != null && isSerdeJson(index, config, channelName, serializer, false)) {
+                    annotationAcceptor.accept(annotation, value);
+                }
+            });
+        }
+    }
+
+    private boolean isSerdeJson(IndexView index, Config config, String channelName, boolean serializer, boolean isKey) {
+        ConfigValue configValue = config.getConfigValue(getConfigName(channelName, serializer, isKey));
+        if (configValue.getValue() != null) {
+            DotName serdeName = DotName.createSimple(configValue.getValue());
+            return serializer ? isSubclassOfJsonSerializer(index, serdeName) : isSubclassOfJsonDeserializer(index, serdeName);
+        }
+        return false;
+    }
+
+    String getConfigName(String channelName, boolean serializer, boolean isKey) {
+        return "mp.messaging." +
+                (serializer ? "outgoing" : "incoming") + "." +
+                channelName + "." +
+                (isKey ? "key" : "value") + "." +
+                (serializer ? "serializer" : "deserializer");
+    }
+
+    private boolean isSubclassOfJsonSerializer(IndexView index, DotName serializerName) {
+        return isSubclassOf(index, DotNames.OBJECT_MAPPER_SERIALIZER, serializerName) ||
+                isSubclassOf(index, DotNames.JSONB_SERIALIZER, serializerName);
+    }
+
+    private boolean isSubclassOfJsonDeserializer(IndexView index, DotName serializerName) {
+        return isSubclassOf(index, DotNames.OBJECT_MAPPER_DESERIALIZER, serializerName) ||
+                isSubclassOf(index, DotNames.JSONB_DESERIALIZER, serializerName);
+    }
+
+    private boolean isSubclassOf(IndexView index, DotName superclass, DotName expectedType) {
+        if (superclass.equals(expectedType)) {
+            return true;
+        }
+        return index.getKnownDirectSubclasses(superclass)
+                .stream()
+                .anyMatch(ci -> ci.name().equals(expectedType));
     }
 }

--- a/extensions/smallrye-reactive-messaging-kafka/deployment/src/test/java/io/quarkus/smallrye/reactivemessaging/kafka/deployment/DefaultSerdeConfigTest.java
+++ b/extensions/smallrye-reactive-messaging-kafka/deployment/src/test/java/io/quarkus/smallrye/reactivemessaging/kafka/deployment/DefaultSerdeConfigTest.java
@@ -810,10 +810,10 @@ public class DefaultSerdeConfigTest {
         doTest(expectations, JacksonDto.class, JacksonDtoDeserializer.class, JacksonDtoInVertxJsonObjectOut.class);
     }
 
-    private static class JacksonDto {
+    static class JacksonDto {
     }
 
-    private static class JacksonDtoDeserializer extends ObjectMapperDeserializer<JacksonDto> {
+    static class JacksonDtoDeserializer extends ObjectMapperDeserializer<JacksonDto> {
         public JacksonDtoDeserializer() {
             super(JacksonDto.class);
         }
@@ -1150,10 +1150,10 @@ public class DefaultSerdeConfigTest {
         doTest(expectations, JsonbDto.class, JsonbDtoSerializer.class, KafkaBytesInJsonbDtoOut.class);
     }
 
-    private static class JsonbDto {
+    static class JsonbDto {
     }
 
-    private static class JsonbDtoSerializer extends JsonbSerializer<JsonbDto> {
+    static class JsonbDtoSerializer extends JsonbSerializer<JsonbDto> {
         @Override
         public byte[] serialize(String topic, Headers headers, JsonbDto data) {
             return null;

--- a/extensions/smallrye-reactive-messaging-kafka/deployment/src/test/java/io/quarkus/smallrye/reactivemessaging/kafka/deployment/ReflectiveClassForValueSerializerPayloadTest.java
+++ b/extensions/smallrye-reactive-messaging-kafka/deployment/src/test/java/io/quarkus/smallrye/reactivemessaging/kafka/deployment/ReflectiveClassForValueSerializerPayloadTest.java
@@ -1,0 +1,335 @@
+package io.quarkus.smallrye.reactivemessaging.kafka.deployment;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.groups.Tuple.tuple;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.HashMap;
+import java.util.Map;
+
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.common.serialization.StringSerializer;
+import org.assertj.core.groups.Tuple;
+import org.eclipse.microprofile.config.Config;
+import org.eclipse.microprofile.reactive.messaging.Channel;
+import org.eclipse.microprofile.reactive.messaging.Emitter;
+import org.eclipse.microprofile.reactive.messaging.Incoming;
+import org.eclipse.microprofile.reactive.messaging.Message;
+import org.eclipse.microprofile.reactive.messaging.Outgoing;
+import org.jboss.jandex.AnnotationInstance;
+import org.jboss.jandex.IndexView;
+import org.jboss.jandex.Indexer;
+import org.jboss.jandex.Type;
+import org.junit.jupiter.api.Test;
+import org.reactivestreams.Publisher;
+
+import io.quarkus.kafka.client.serialization.JsonbDeserializer;
+import io.quarkus.kafka.client.serialization.ObjectMapperSerializer;
+import io.quarkus.smallrye.reactivemessaging.kafka.deployment.DefaultSerdeConfigTest.JacksonDto;
+import io.quarkus.smallrye.reactivemessaging.kafka.deployment.DefaultSerdeConfigTest.JacksonDtoDeserializer;
+import io.quarkus.smallrye.reactivemessaging.kafka.deployment.DefaultSerdeConfigTest.JsonbDto;
+import io.quarkus.smallrye.reactivemessaging.kafka.deployment.DefaultSerdeConfigTest.JsonbDtoSerializer;
+import io.smallrye.config.SmallRyeConfigBuilder;
+import io.smallrye.config.common.MapBackedConfigSource;
+import io.smallrye.mutiny.Multi;
+import io.smallrye.reactive.messaging.MutinyEmitter;
+import io.smallrye.reactive.messaging.kafka.KafkaRecord;
+import io.smallrye.reactive.messaging.kafka.Record;
+
+class ReflectiveClassForValueSerializerPayloadTest {
+    private static void doTest(Tuple[] expectations, Map<String, String> configs, Class<?>... classesToIndex) {
+        Map<AnnotationInstance, Type> annotations = new HashMap<>();
+
+        IndexView index = index(classesToIndex);
+        Config config = configFromMap(configs);
+        SmallRyeReactiveMessagingKafkaProcessor processor = new SmallRyeReactiveMessagingKafkaProcessor();
+        processor.processOutgoingForReflectiveClassPayload(index, config, annotations::put);
+        processor.processOutgoingChannelForReflectiveClassPayload(index, config, annotations::put);
+        processor.processIncomingForReflectiveClassPayload(index, config, annotations::put);
+        processor.processIncomingChannelForReflectiveClassPayload(index, config, annotations::put);
+
+        assertThat(annotations)
+                .extractingFromEntries(e -> e.getKey().value().asString(), e -> e.getValue().name().toString())
+                .containsOnly(expectations);
+    }
+
+    private static class JacksonDtoSerializer extends ObjectMapperSerializer<JacksonDto> {
+    }
+
+    private static class JacksonDtoSerde {
+
+        // Emitter @Channel
+        @Channel("outgoing-channel1")
+        Emitter<JacksonDto> emitter1;
+
+        @Channel("outgoing-channel2")
+        MutinyEmitter<JacksonDto> emitter2;
+
+        @Channel("outgoing-channel3")
+        MutinyEmitter<JacksonDto> emitter3;
+
+        @Channel("outgoing-channel4")
+        Publisher<JacksonDto> notExpected1;
+
+        @Channel("outgoing-channel5")
+        Emitter<JacksonDto> notExpected2;
+
+        // @Outgoing
+        @Outgoing("outgoing1")
+        Publisher<Message<JacksonDto>> outgoing1() {
+            return null;
+        }
+
+        @Outgoing("outgoing2")
+        Message<JacksonDto> outgoing2() {
+            return null;
+        }
+
+        @Outgoing("outgoing3")
+        Publisher<JacksonDto> outgoing3() {
+            return null;
+        }
+
+        @Outgoing("outgoing4")
+        Multi<Record<String, JacksonDto>> outgoing4() {
+            return null;
+        }
+
+        @Outgoing("outgoing5")
+        JacksonDto outgoing5() {
+            return null;
+        }
+
+        @Outgoing("outgoing6")
+        JacksonDto notExpected() {
+            return null;
+        }
+
+        // @Incoming
+        @Incoming("incoming1")
+        void incoming1(Message<JacksonDto> param) {
+        }
+
+        @Incoming("incoming2")
+        void incoming2(Record<String, JacksonDto> param) {
+        }
+
+        @Incoming("incoming3")
+        void incoming3(JacksonDto param) {
+        }
+
+        @Incoming("incoming4")
+        void incoming4(ConsumerRecord<JacksonDto, String> param) {
+        }
+
+        @Incoming("incoming5")
+        void notExpected(Publisher<JacksonDto> param) {
+        }
+
+        // Incoming @Channel
+        @Channel("incoming-channel1")
+        Multi<JacksonDto> incoming1;
+
+        @Channel("incoming-channel2")
+        Publisher<JacksonDto> incoming2;
+
+        @Channel("incoming-channel3")
+        Multi<Message<JacksonDto>> incoming3;
+
+        @Channel("incoming-channel4")
+        Publisher<Record<JacksonDto, String>> incoming4;
+
+    }
+
+    @Test
+    void processAnnotationsForReflectiveClassJacksonPayload() {
+        Tuple[] expectations = {
+                tuple("outgoing-channel1", JacksonDto.class.getName()),
+                tuple("outgoing-channel2", JacksonDto.class.getName()),
+                tuple("outgoing-channel3", JacksonDto.class.getName()),
+                tuple("outgoing1", JacksonDto.class.getName()),
+                tuple("outgoing2", JacksonDto.class.getName()),
+                tuple("outgoing3", JacksonDto.class.getName()),
+                tuple("outgoing4", JacksonDto.class.getName()),
+                tuple("outgoing5", JacksonDto.class.getName()),
+                tuple("incoming1", JacksonDto.class.getName()),
+                tuple("incoming2", JacksonDto.class.getName()),
+                tuple("incoming3", JacksonDto.class.getName()),
+                tuple("incoming4", JacksonDto.class.getName()),
+                tuple("incoming-channel1", JacksonDto.class.getName()),
+                tuple("incoming-channel2", JacksonDto.class.getName()),
+                tuple("incoming-channel3", JacksonDto.class.getName()),
+                tuple("incoming-channel4", JacksonDto.class.getName()),
+        };
+        Map<String, String> configMap = Map.ofEntries(
+                Map.entry("mp.messaging.outgoing.outgoing-channel1.value.serializer", JacksonDtoSerializer.class.getName()),
+                Map.entry("mp.messaging.outgoing.outgoing-channel2.value.serializer", JacksonDtoSerializer.class.getName()),
+                Map.entry("mp.messaging.outgoing.outgoing-channel3.value.serializer", JacksonDtoSerializer.class.getName()),
+                Map.entry("mp.messaging.outgoing.outgoing-channel4.value.serializer", JacksonDtoSerializer.class.getName()),
+                Map.entry("mp.messaging.outgoing.outgoing1.value.serializer", JacksonDtoSerializer.class.getName()),
+                Map.entry("mp.messaging.outgoing.outgoing2.value.serializer", JacksonDtoSerializer.class.getName()),
+                Map.entry("mp.messaging.outgoing.outgoing3.value.serializer", ObjectMapperSerializer.class.getName()),
+                Map.entry("mp.messaging.outgoing.outgoing4.key.serializer", StringSerializer.class.getName()),
+                Map.entry("mp.messaging.outgoing.outgoing4.value.serializer", JacksonDtoSerializer.class.getName()),
+                Map.entry("mp.messaging.outgoing.outgoing5.value.serializer", JacksonDtoSerializer.class.getName()),
+                Map.entry("mp.messaging.incoming.incoming1.value.deserializer", JacksonDtoDeserializer.class.getName()),
+                Map.entry("mp.messaging.incoming.incoming2.value.deserializer", JacksonDtoDeserializer.class.getName()),
+                Map.entry("mp.messaging.incoming.incoming3.value.deserializer", JacksonDtoDeserializer.class.getName()),
+                Map.entry("mp.messaging.incoming.incoming4.key.deserializer", JacksonDtoDeserializer.class.getName()),
+                Map.entry("mp.messaging.incoming.incoming-channel1.value.deserializer", JacksonDtoDeserializer.class.getName()),
+                Map.entry("mp.messaging.incoming.incoming-channel2.value.deserializer", JacksonDtoDeserializer.class.getName()),
+                Map.entry("mp.messaging.incoming.incoming-channel3.value.deserializer", JacksonDtoDeserializer.class.getName()),
+                Map.entry("mp.messaging.incoming.incoming-channel4.key.deserializer", JacksonDtoDeserializer.class.getName()));
+
+        doTest(expectations, configMap, JacksonDto.class, JacksonDtoSerializer.class, JacksonDtoDeserializer.class,
+                JacksonDtoSerde.class);
+    }
+
+    static class JsonbDtoDeserializer extends JsonbDeserializer<JsonbDto> {
+
+        public JsonbDtoDeserializer() {
+            super(JsonbDto.class);
+        }
+    }
+
+    private static class JsonbDtoSerde {
+        // Emitter @Channel
+        @Channel("outgoing-channel1")
+        Emitter<Record<String, JsonbDto>> emitter1;
+
+        @Channel("outgoing-channel2")
+        MutinyEmitter<JsonbDto> emitter2;
+
+        @Channel("outgoing-channel3")
+        MutinyEmitter<KafkaRecord<JsonbDto, Long>> emitter3;
+
+        @Channel("outgoing-channel4")
+        Emitter<JsonbDto> emitter4;
+
+        // @Outgoing
+        @Outgoing("outgoing1")
+        Publisher<Message<JsonbDto>> outgoing1() {
+            return null;
+        }
+
+        @Outgoing("outgoing2")
+        Message<JsonbDto> outgoing2() {
+            return null;
+        }
+
+        @Outgoing("outgoing3")
+        Publisher<JsonbDto> outgoing3() {
+            return null;
+        }
+
+        @Outgoing("outgoing4")
+        Multi<Record<String, JsonbDto>> outgoing4() {
+            return null;
+        }
+
+        @Outgoing("outgoing5")
+        JsonbDto outgoing5() {
+            return null;
+        }
+
+        // @Incoming
+        @Incoming("incoming1")
+        void incoming1(Message<JsonbDto> param) {
+        }
+
+        @Incoming("incoming2")
+        void incoming2(Record<String, JsonbDto> param) {
+        }
+
+        @Incoming("incoming3")
+        void incoming3(JsonbDto param) {
+        }
+
+        @Incoming("incoming4")
+        void incoming4(ConsumerRecord<JsonbDto, String> param) {
+        }
+
+        // Incoming @Channel
+        @Channel("incoming-channel1")
+        Multi<JsonbDto> incoming1;
+
+        @Channel("incoming-channel2")
+        Publisher<JsonbDto> incoming2;
+
+        @Channel("incoming-channel3")
+        Multi<Message<JsonbDto>> incoming3;
+
+        @Channel("incoming-channel4")
+        Publisher<Record<JsonbDto, String>> incoming4;
+
+    }
+
+    @Test
+    void processAnnotationsForReflectiveClassJsonbPayload() {
+        Tuple[] expectations = {
+                tuple("outgoing-channel1", JsonbDto.class.getName()),
+                tuple("outgoing-channel2", JsonbDto.class.getName()),
+                tuple("outgoing-channel3", JsonbDto.class.getName()),
+                tuple("outgoing-channel4", JsonbDto.class.getName()),
+                tuple("outgoing1", JsonbDto.class.getName()),
+                tuple("outgoing2", JsonbDto.class.getName()),
+                tuple("outgoing3", JsonbDto.class.getName()),
+                tuple("outgoing4", JsonbDto.class.getName()),
+                tuple("outgoing5", JsonbDto.class.getName()),
+                tuple("incoming1", JsonbDto.class.getName()),
+                tuple("incoming2", JsonbDto.class.getName()),
+                tuple("incoming3", JsonbDto.class.getName()),
+                tuple("incoming4", JsonbDto.class.getName()),
+                tuple("incoming-channel1", JsonbDto.class.getName()),
+                tuple("incoming-channel2", JsonbDto.class.getName()),
+                tuple("incoming-channel3", JsonbDto.class.getName()),
+                tuple("incoming-channel4", JsonbDto.class.getName()),
+        };
+
+        Map<String, String> configMap = Map.ofEntries(
+                Map.entry("mp.messaging.outgoing.outgoing-channel1.value.serializer", JsonbDtoSerializer.class.getName()),
+                Map.entry("mp.messaging.outgoing.outgoing-channel2.value.serializer", JsonbDtoSerializer.class.getName()),
+                Map.entry("mp.messaging.outgoing.outgoing-channel3.key.serializer", JsonbDtoSerializer.class.getName()),
+                Map.entry("mp.messaging.outgoing.outgoing-channel4.value.serializer", JsonbDtoSerializer.class.getName()),
+                Map.entry("mp.messaging.outgoing.outgoing1.value.serializer", JsonbDtoSerializer.class.getName()),
+                Map.entry("mp.messaging.outgoing.outgoing2.value.serializer", JsonbDtoSerializer.class.getName()),
+                Map.entry("mp.messaging.outgoing.outgoing3.value.serializer", JsonbDtoSerializer.class.getName()),
+                Map.entry("mp.messaging.outgoing.outgoing4.value.serializer", JsonbDtoSerializer.class.getName()),
+                Map.entry("mp.messaging.outgoing.outgoing5.value.serializer", JsonbDtoSerializer.class.getName()),
+
+                Map.entry("mp.messaging.incoming.incoming1.value.deserializer", JsonbDtoDeserializer.class.getName()),
+                Map.entry("mp.messaging.incoming.incoming2.value.deserializer", JsonbDtoDeserializer.class.getName()),
+                Map.entry("mp.messaging.incoming.incoming3.value.deserializer", JsonbDtoDeserializer.class.getName()),
+                Map.entry("mp.messaging.incoming.incoming4.key.deserializer", JsonbDtoDeserializer.class.getName()),
+                Map.entry("mp.messaging.incoming.incoming-channel1.value.deserializer", JsonbDtoDeserializer.class.getName()),
+                Map.entry("mp.messaging.incoming.incoming-channel2.value.deserializer", JsonbDtoDeserializer.class.getName()),
+                Map.entry("mp.messaging.incoming.incoming-channel3.value.deserializer", JsonbDtoDeserializer.class.getName()),
+                Map.entry("mp.messaging.incoming.incoming-channel4.key.deserializer", JsonbDtoDeserializer.class.getName()));
+
+        doTest(expectations, configMap, JsonbDto.class, JsonbDtoSerializer.class, JsonbDtoDeserializer.class,
+                JsonbDtoSerde.class);
+    }
+
+    private static IndexView index(Class<?>... classes) {
+        Indexer indexer = new Indexer();
+        for (Class<?> clazz : classes) {
+            try {
+                try (InputStream stream = DefaultSerdeConfigTest.class.getClassLoader()
+                        .getResourceAsStream(clazz.getName().replace('.', '/') + ".class")) {
+                    indexer.index(stream);
+                }
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+        }
+        return indexer.complete();
+    }
+
+    private static Config configFromMap(Map<String, String> configMap) {
+        return new SmallRyeConfigBuilder()
+                .withSources(new MapBackedConfigSource("config-map", configMap) {
+                })
+                .build();
+    }
+}

